core/src/main/java/org/sonarsource/sonarlint/core/tracking/SecurityHotspotMatchingServiceImpl.java

```diff
@@ -0,0 +1,165 @@
+import org.sonarsource.sonarlint.core.clientapi.backend.tracking.MatchWithServerSecurityHotspotsParams;
+import org.sonarsource.sonarlint.core.clientapi.backend.tracking.MatchWithServerSecurityHotspotsResponse;
+import org.sonarsource.sonarlint.core.clientapi.backend.tracking.SecurityHotspotMatchingService;
+import org.sonarsource.sonarlint.core.clientapi.backend.tracking.ServerMatchedSecurityHotspotDto;
+import org.sonarsource.sonarlint.core.commons.Binding;
+import org.sonarsource.sonarlint.core.commons.log.SonarLintLogger;
+import org.sonarsource.sonarlint.core.issuetracking.Trackable;
+import org.sonarsource.sonarlint.core.issuetracking.Tracker;
+import org.sonarsource.sonarlint.core.repository.config.ConfigurationRepository;
+import org.sonarsource.sonarlint.core.repository.vcs.ActiveSonarProjectBranchRepository;
+import org.sonarsource.sonarlint.core.serverapi.hotspot.ServerHotspot;
+import org.sonarsource.sonarlint.core.serverconnection.StorageService;
+import org.sonarsource.sonarlint.core.sync.SynchronizationServiceImpl;
+
+import static org.sonarsource.sonarlint.core.utils.FutureUtils.waitForTask;
+import static org.sonarsource.sonarlint.core.utils.FutureUtils.waitForTasks;
+
+@Named
+@Singleton
+public class SecurityHotspotMatchingServiceImpl implements SecurityHotspotMatchingService {
+  private static final int FETCH_ALL_SECURITY_HOTSPOTS_THRESHOLD = 10;
+  private static final SonarLintLogger LOG = SonarLintLogger.get();
+  private final ConfigurationRepository configurationRepository;
+  private final StorageService storageService;
+  private final ActiveSonarProjectBranchRepository activeSonarProjectBranchRepository;
+  private final SynchronizationServiceImpl synchronizationService;
+  private final ExecutorService executorService;
+
+  public SecurityHotspotMatchingServiceImpl(ConfigurationRepository configurationRepository, StorageService storageService,
+    ActiveSonarProjectBranchRepository activeSonarProjectBranchRepository, SynchronizationServiceImpl synchronizationService) {
+    this.configurationRepository = configurationRepository;
+    this.storageService = storageService;
+    this.activeSonarProjectBranchRepository = activeSonarProjectBranchRepository;
+    this.synchronizationService = synchronizationService;
+    this.executorService = Executors.newSingleThreadExecutor(r -> new Thread(r, "sonarlint-server-tracking-hotspot-updater"));
+  }
+
+  @Override
+  public CompletableFuture<MatchWithServerSecurityHotspotsResponse> matchWithServerSecurityHotspots(MatchWithServerSecurityHotspotsParams params) {
+    return CompletableFutures.computeAsync(cancelChecker -> {
+      var configurationScopeId = params.getConfigurationScopeId();
+      var effectiveBindingOpt = configurationRepository.getEffectiveBinding(configurationScopeId);
+      var activeBranchOpt = activeSonarProjectBranchRepository.getActiveSonarProjectBranch(configurationScopeId);
+      if (effectiveBindingOpt.isEmpty() || activeBranchOpt.isEmpty()) {
```

Comment:
May add a debug log here to know we fall in this case.

Reply:
This method will also be called in standalone mode, so I'm not sure we should print a trace at the end of each analysis, it's already quite verbose